File: bindings/nodejs/test/auth.test.js
'use strict';

/**
 * `crypto_auth` - three categories per D-64/D-65: correctness (round trip), rejection (tampered
 * message, wrong key), misuse (wrong-length key/tag - foreclosed at the Rust layer by fixed-size
 * arrays, D-66, so `auth` itself is infallible; only the JS-boundary length checks are testable
 * here). Mirrors bindings/python/tests/test_auth.py.
 */

const test = require('node:test');
const assert = require('node:assert/strict');
const dstu = require('../js/index.js');

test('auth/verify round-trips', () => {
  const key = dstu.authKeygen();
  const message = Buffer.from('a message both parties want to confirm is unmodified');
  const tag = dstu.auth(key, message);
  dstu.authVerify(key, message, tag);
});

test('tampered message is rejected', () => {
  const key = dstu.authKeygen();
  const tag = dstu.auth(key, Buffer.from('original message'));
  assert.throws(() => dstu.authVerify(key, Buffer.from('a different message'), tag));
});

test('wrong key is rejected', () => {
  const key = dstu.authKeygen();
  const otherKey = dstu.authKeygen();
  const tag = dstu.auth(key, Buffer.from('message'));
  assert.throws(() => dstu.authVerify(otherKey, Buffer.from('message'), tag));
});

test('wrong-length key is rejected', () => {
  assert.throws(() => dstu.auth(Buffer.from('too short'), Buffer.from('message')));
});

test('wrong-length tag is rejected', () => {
  const key = dstu.authKeygen();
  assert.throws(() => dstu.authVerify(key, Buffer.from('message'), Buffer.from('too short')));
});

File: bindings/java/src/test/java/ua/dstucrypto/dstucore/AuthTest.java
package ua.dstucrypto.dstucore;

import org.junit.jupiter.api.Test;

import java.io.UnsupportedEncodingException;

import static org.junit.jupiter.api.Assertions.assertThrows;

/**
 * {@code crypto_auth} - three categories per D-64/D-65: correctness (round trip), rejection
 * (tampered message, wrong key), misuse (wrong-length key/tag).
 */
class AuthTest {
    private static byte[] bytes(String s) throws UnsupportedEncodingException {
        return s.getBytes("UTF-8");
    }

    @Test
    void authVerifyRoundTrips() throws Exception {
        byte[] key = Auth.keygen();
        byte[] tag = Auth.auth(key, bytes("a message both parties want to confirm is unmodified"));
        Auth.verify(key, bytes("a message both parties want to confirm is unmodified"), tag);
    }

    @Test
    void tamperedMessageIsRejected() throws Exception {
        byte[] key = Auth.keygen();
        byte[] tag = Auth.auth(key, bytes("original message"));
        assertThrows(DstuException.class, () -> Auth.verify(key, bytes("a different message"), tag));
    }

    @Test
    void wrongKeyIsRejected() throws Exception {
        byte[] key = Auth.keygen();
        byte[] otherKey = Auth.keygen();
        byte[] tag = Auth.auth(key, bytes("message"));
        assertThrows(DstuException.class, () -> Auth.verify(otherKey, bytes("message"), tag));
    }

    @Test
    void wrongLengthKeyIsRejected() throws Exception {
        byte[] tooShort = bytes("too short");
        assertThrows(IllegalArgumentException.class, () -> Auth.auth(tooShort, bytes("message")));
    }

    @Test
    void wrongLengthTagIsRejected() throws Exception {
        byte[] key = Auth.keygen();
        byte[] tooShort = bytes("too short");
        assertThrows(IllegalArgumentException.class, () -> Auth.verify(key, bytes("message"), tooShort));
    }
}
